### PR TITLE
Support for built-in @type and @spec

### PR DESCRIPTION
**Pull request by orsinium**
The complete version of #109

Adds the following public functions:

1. `fetch_spec` to extract type from `@spec` of the given function.
1. `fetch_type` to extract type from `@type` with the given name.
1. `apply` which is the same as `Kernel.apply` but also checks types (both arguments and result).
1. `ensure_types` macro which extracts types from `@spec` and puts it into `apply`.

The last one is experimental and isn't very stable. Areas for improvement:

1. If it can't fetch the given spec, it will fail with a pattern matching error. It would be better to provide a friendlier error for such cases.
2. I've created The full function name inference (`ast_to_mfa`) using only my own head and my own experiments in REPL. It might miss some corner cases, I'm not proficient with Elixir AST yet.

Other than that, it should be good enough. The parser supports all types that TypeCheck does and even more. It also correctly infers all remote types in `@spec`.

### CODE REVIEW

**Comment by orsinium** [comment]
In theory, we also could use `Code.Typespec.spec_to_quoted` for `convert`, but custom implementation gives more control.

File: lib/type_check.ex
Comment (Qqwy) [inline lib/type_check.ex:348]: On `{:error, problem}` being returned from `fetch_spec` you might want to raise a `TypeCheck.CompileError, problem`.

File: lib/type_check/internals/parser.ex
Comment (Qqwy) [inline lib/type_check/internals/parser.ex:121]: I'm not entirely convinced that we should always be optimistic rather than being explicit about what we do and don't support.

Maybe we should add an option for the user to decide between them.
Something for a later version; please remind me that we add an issue for this.

File: lib/type_check/internals/parser.ex
Comment (Qqwy) [inline lib/type_check/internals/parser.ex:168]: Support for references and ports has been added, so this can be fixed :muscle: .

File: lib/type_check/internals/parser.ex
Comment (Qqwy) [inline lib/type_check/internals/parser.ex:222]: They are, but there is a good reason for them (c.f. "IO lists"). However, TypeCheck itself also does not properly support them yet. Defaulting to lists definitely is better than defaulting to any here.

However, we will need to revisit this in the future. Please remind me (with an issue?) once this PR is merged.

File: test/type_check/builtin_test.exs
Comment (Qqwy) [inline test/type_check/builtin_test.exs:142]: Very nice :+1:

File: test/type_check/internals/parser_test.exs
Comment (Qqwy) [inline test/type_check/internals/parser_test.exs:12]: I love how much testing you're able to do with this little code!
Hooray for reflection :blush:

File: test/type_check/internals/parser_test.exs
Comment (Qqwy) [inline test/type_check/internals/parser_test.exs:86]: I tip my hat at your test-writing skills! :heart_eyes:

**Review by Qqwy — CHANGES_REQUESTED** [review]
## Stability

I think it would be a good idea to move `ensure_types!`/`apply`/`fetch_spec` and `fetch_type` to a separate module, say `TypeCheck.Experimental` to make it even more clear that their API and internals might be subject to change.
(Side note: I still hope to release v1.0 of this library around ElixirConf EU, which is why we should be very clear which parts are and are not yet stable.)

Obviously the 'Experimental' doc sections might then be moved to the moduledoc.

## Raising/non-raising

I'm not 100% certain (would love to know your opinion) but currently leaning towards also providing non-raising (error-tuple returning) versions of at least `apply!` to allow error-recovery without exceptions.


---

Please don't feel like all my comments have to be fixed *right now*. A lot of them need to be addressed at some point, but they might wait until after this PR is merged.


Thank you _so much_ for all your hard work on this! 
It is absolutely amazing. :heart_on_fire:

File: lib/type_check.ex
Comment (orsinium) [reply lib/type_check.ex:345]: We ensure that the values have the correct types, hence the name is "ensure types". Or you can think about it as type-specific "ensure" from DbC. I'm open to better ideas but "well-typed" seems also confusing. What is even "well-typed"? Sounds like the annotations satisfy some expected type coverage or whatever.

File: lib/type_check.ex
Comment (orsinium) [reply lib/type_check.ex:350]: It failed for me with "function :erl_eval.rtype/0 is undefined or private", so I decided not to overthink it for now. Make it work first, and optimize it later. The API won't change anyway.

File: lib/type_check.ex
Comment (orsinium) [reply lib/type_check.ex:356]: I think it needs careful thinking. What if the wrapped function returns `{:ok, result}`? Would you unwrap it or return `{:ok, {:ok, result}}`? Also, I don't see just yet an easy way to do it without much duplication, so I'd better first establish internals and only then copy-paste it for a monadic result.

File: lib/type_check/internals/parser.ex
Comment (orsinium) [reply lib/type_check/internals/parser.ex:24]: It is used in tests, I pass the bytecode of generated module there as an input. The public wrapper accepts only `module()`, though.

File: lib/type_check/internals/parser.ex
Comment (orsinium) [reply lib/type_check/internals/parser.ex:121]: That's why I've added `ctx.default`. I also thought about adding a `ctx.raise` flag. Neither is exposed to the user just yet, though.

**Comment by orsinium** [comment]
> I think it would be a good idea to move `ensure_types!/apply/fetch_spec` and fetch_type to a separate module, say TypeCheck.Experimental to make it even more clear that their API and internals might be subject to change.

I don't think their API will change in a breaking manner. Do you? At least, I'm pretty sure about the stability of `fetch_spec` and `fetch_type`. They might later support some options but the contract "Give me MFA and I'll return TypeCheck type" will stay forever unchanged.

File: lib/type_check.ex
Comment (Qqwy) [reply lib/type_check.ex:360]: Yes, it would return `{:ok, {:ok, result}}`.
But yes, we should wait until after this PR is merged and the internals are established.

File: lib/type_check.ex
Comment (Qqwy) [reply lib/type_check.ex]: :thinking: You are right. 'Well-Typed' did not mean what I remembered it to mean. It actually means "a semantically correct type" (the inverse of 'ill-typed') rather than "a value conforming to its type".

I still think we should give it another name as `ensure_types!` makes it sound like we try to ensure that the input _is_ a type rather than _an expression adhering_ to its type.
When I think of a better name, I'll let you know.

**Comment by Qqwy** [comment]
> I don't think their API will change in a breaking manner. Do you?

- The name for `ensure_types!` will probably still change
- `fetch_spec`/ `fetch_type` 
  - As written right now, might change what types they return for a particular type/spec as anything unrecognized is currently turned into any. 
- I'm not entirely sure where these functions should live. Probably _not_ in the main `TypeCheck` module, as that module is meant only for functions that see very frequent usage. Maybe `ensure_types!` (whatever it might be called later) and `apply!` might be included there, `fetch_spec` might make more sense in the `TypeCheck.Spec` module and `fetch_type` in the `TypeCheck.Type` module. On the other hand, it also makes sense to group all functions that deal with pre-existing specs/types together in a single module. :thinking: 

Because of these considerations, I think it is better to put them in a module that is very clear about not being fully stable yet for the time being, so that we can defer these choices until later, rather than finding out we chose badly and being unable to change it in a backwards-compatible way.

File: lib/type_check.ex
Comment (Qqwy) [reply lib/type_check.ex]: Maybe `enforce_types!`?

**Comment by orsinium** [comment]
Good idea. I propose to call the module something like "TypeCheck.Classic" and say in its docs that the module is experimental.

**Comment by Qqwy** [comment]
> Good idea. I propose to call the module something like "TypeCheck.Classic" and say in its docs that the module is experimental.

What about the name `TypeCheck.External`? 'Classic' has the connotation of referring to an old version of the code, which I do not think is appropriate here.

'External' indicates that you can use it to work with types of modules which you do not directly have access to yourself.

**Comment by orsinium** [comment]
Changes since the last review:

1. Support for recursive types, with recursion depth limit
2. Support for user types
3. Moved some funcs into `TypeCheck.External`
4. Added `TypeCheck.apply` to return a monad
5. Support for reference, port, maybe_improper_list
6. More tests
7. Simplification of `type | any` to just `any` (with any number of types inside).
8. Simplification of nested union for any number of arguments.

**Comment by orsinium** [comment]
A few more changes:

1. Support for nonempty_maybe_improper_list because #120 is merged.
2. Moved `apply` and `apply!` into `External` since you need them only with "external", non-type-check-annotated, functions.

File: lib/type_check/external.ex
Comment (Qqwy) [inline lib/type_check/external.ex:137]: If you want, you can change this to re-use `apply`, by writing:

```
case apply(type, module, function, args) do
  {:ok, result} -> result
  {:error, exception} -> raise exception
end
```

File: lib/type_check/internals/parser.ex
Comment (Qqwy) [inline lib/type_check/internals/parser.ex:2]: We do not want this module to show up in the generated public documentation, as this module is not intended for public usage but only for implementers,
so write `@moduledoc false` and move the current documentation to comments right below it.

File: test/type_check/type_error/formatter_test.exs
Comment (Qqwy) [inline test/type_check/type_error/formatter_test.exs:41]: Thank you for making this more clear! :green_heart:

**Review by Qqwy — APPROVED** [review]
The PR is shaping up nicely! :blush: 
There are only some tiny nitpicks left, most of which are optional.

File: lib/type_check/external.ex
Comment (orsinium) [reply lib/type_check/external.ex:183]: It's not the same as `check_params([], [])`.

1. If there are more types than passed arguments, it's possible that some function arguments have a default.
2. If there are more arguments passed than we have types, it's most likely an error but I pass it through, so BEAM can try to call the function and fail.

So, we need to check here that **at least one** list is empty, not necessarily both. You can replace it with 2 clauses `check_params([], _)` and `check_params(_, [])` but why do the same thing twice?

File: lib/type_check/external.ex
Comment (orsinium) [reply lib/type_check/external.ex:137]: I didn't want to mess with the stacktrace, TBH. Otherwise, yes, just one function call is actually different.

File: test/type_check/type_error/formatter_test.exs
Comment (orsinium) [reply test/type_check/type_error/formatter_test.exs:41]: It's not just clarity, I've added all new modules here that were missed and commented out those of them that were failing. The rest is just `mix format`.

**Review by Qqwy — APPROVED** [review]
Wonderful work! :heart_on_fire:

**Pull request merged**
